### PR TITLE
Introduce batching logic in sync_task

### LINKED ISSUE #220: [Proposal] Improving Xline throughput in the high-load scenario
# Motivation
Improving Xline throughput in the high-load scenario.

# Implementation
There are three questions we should answer before we start discussing batching mechanism.

1. What does a batch look like?
    The so-called batch we mentioned above is an AppenEntries request which contains several `LogEntry` while every LogEntry has only one command. 

2. How to dispatch a batch？
    When one of the two following conditions is satisfied, a batch will be dispatched:
    1. When the size of a batch reaches its limit. 
       gRPC has imposed a 4MB cap on the size of a message. And in some raft implementations, like LogCabin, this message size cap is about 1MB([MAX_MESSAGE_LENGTH](https://github.com/logcabin/logcabin/blob/ee6c55ae9744b82b451becd9707d26c7c1b6bbfb/Protocol/Common.h#L78)). A  way better than hardcoding this limitation is to make it configurable.

    2. When the batching_timeout is exceeded. 
        Since we are using AppenEntries to optimize the Heartbeat, batching_timeout should not exceed the heartbeat_interval(300ms). Currently, the latency for a single request does not exceed 10ms. Therefore, a reasonable duration of batching_timeout could be from 15ms to 25ms. Same as batching_max_size, we can also make batching_timeout configurable. 

3. Who is in charge of collecting a batch and dispatching it？
    Currently, sync_task is responsible for replicating log entries, so we can place the  batch-collecting logic and dispatch strategy here. 

# Extra discussion
Q: How to retry if a batch dispatch failed?
A: If a batch dispatch failed, it will fall back to the calibrate and retry in there.

Q: What would happen if leadership changed during batch dispatching? 
A： The possible cases of leadership change are:
1. The original leader crashed:
  1. The original leader restarts immediately and becomes the new leader. It will execute the calibration process
  2. A follower will initiate an election and win the election finally, resulting in an increase in the current term. Since the unsent batch must not commit, the new leader can truncate the log in its calibration process. It does not violate the safety principle of the consensus algorithm.
  5. When a network partition occurs, the original leader enters the minority part, and the majority part elects a new leader. The original leader can send its batching log as usual, but these AppenEntries requests will never be accepted. Once the network partition is restored, the original leader will step down as a follower, and the new leader will truncate its log in the calibration process.

### PR DESCRIPTION
**Pull request by Phoenix500526**
Please briefly answer these questions:
Improving Xline throughput by batching in the high-load scenario. This pr is based on #217.  Close #220.

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
Improving Xline throughput by batching in the high-load scenario.

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)
No

### CODE REVIEW

File: curp/src/server/curp_node.rs [hunk 1]
```diff
@@ -225,7 +233,7 @@ impl<C: 'static + Command> CurpNode<C> {
         curp_cfg: Arc<CurpConfig>,
         tx_filter: Option<Box<dyn TxFilter>>,
     ) -> Result<Self, CurpError> {
-        let (sync_tx, sync_rx) = mpsc::unbounded_channel();
+        let (sync_tx, sync_rx) = flume::unbounded();
```
Comment (rogercloud) [inline curp/src/server/curp_node.rs:236]: what's the difference of these two unbounded channels?

File: curp/src/server/curp_node.rs [hunk 2]
```diff
@@ -490,49 +498,92 @@ impl<C: 'static + Command> CurpNode<C> {
...
-    /// Sync task is responsible for replicating log entries
-    async fn sync_task(
-        curp: Arc<RawCurp<C>>,
-        connects: HashMap<ServerId, Arc<impl ConnectApi>>,
-        mut sync_rx: mpsc::UnboundedReceiver<LogIndex>,
+    /// invoke `send_log_to_follower` for every follower
+    fn bcast_append_entries(
+        curp: &Arc<RawCurp<C>>,
+        connects: &HashMap<ServerId, Arc<impl ConnectApi>>,
+        batch: Vec<LogIndex>,
```
Comment (rogercloud) [inline curp/src/server/curp_node.rs:509]: If the batch's contents are always the continuous number, can we do it better? For instance the first log index and the length.

File: curp/src/server/curp_node.rs [hunk 3]
```diff
@@ -490,49 +498,92 @@ impl<C: 'static + Command> CurpNode<C> {
...
+        curp: Arc<RawCurp<C>>,
+        connects: HashMap<ServerId, Arc<impl ConnectApi>>,
+        sync_rx: flume::Receiver<(LogIndex, usize)>,
+    ) {
+        let batch_timeout = curp.cfg().batch_timeout;
+        let batch_max_size = curp.cfg().batch_max_size;
+        let connects = Arc::new(connects);
+        let timer = sleep(batch_timeout);
+        let batch = Mutex::new(Vec::new());
+        let batch_size = AtomicUsize::new(0);
```
Comment (rogercloud) [inline curp/src/server/curp_node.rs:553]: Why do you use atomic number here? I don't think it'll be accessed parallelly.

File: utils/src/lib.rs [hunk 4]
```diff
@@ -264,6 +264,42 @@ pub fn parse_rotation(s: &str) -> Result<RotationConfig, ConfigParseError> {
     }
 }
 
+/// Parse bytes from string
+/// # Errors
+/// Return error when parsing the given string to usize failed
+#[inline]
+#[allow(clippy::integer_arithmetic)]
+pub fn parse_batch_bytes(s: &str) -> Result<usize, ConfigParseError> {
+    if s.ends_with("kb") {
```
Comment (rogercloud) [inline utils/src/lib.rs:273]: let's make it case insensitive.

File: curp/src/server/curp_node.rs (hunk 2 above)
Comment (Phoenix500526) [reply curp/src/server/curp_node.rs:509]: Fixed

File: curp/src/server/curp_node.rs (hunk 3 above)
Comment (Phoenix500526) [reply curp/src/server/curp_node.rs]: Fixed.

File: curp/src/server/raw_curp/mod.rs [hunk 5]
```diff
@@ -683,27 +684,47 @@ impl<C: 'static + Command> RawCurp<C> {
         self.ctx.leader_tx.subscribe()
     }
 
-    /// Get `append_entries` request for log[i]
+    /// Get `append_entries` request for log batch
```
[the remark this answers is not in the document]
Comment (Phoenix500526) [reply curp/src/server/raw_curp/mod.rs:687]: Fixed

File: utils/src/lib.rs (hunk 4 above)
Comment (Phoenix500526) [reply utils/src/lib.rs:275]: Fixed

File: utils/src/lib.rs [hunk 6]
```diff
@@ -264,6 +264,42 @@ pub fn parse_rotation(s: &str) -> Result<RotationConfig, ConfigParseError> {
...
+        if let Some(value) = s.strip_suffix("kb") {
+            Ok(value.parse::<usize>()? * 1024)
+        } else {
+            Err(ConfigParseError::InvalidValue(format!(
+                "the value of size should not be empty. ({s})"
+            )))
+        }
+    } else if s.ends_with("mb") {
```
[the remark this answers is not in the document]
Comment (Phoenix500526) [reply utils/src/lib.rs:283]: Fixed

File: curp/src/server/curp_node.rs (hunk 1 above)
Comment (Phoenix500526) [reply curp/src/server/curp_node.rs]: Fixed

**Comment by Phoenix500526** [comment]
New batching is in pr #227

**Pull request closed without merging**